### ISSUE TITLE
Calculate BlockId from Tendermint Header

### ISSUE THREAD

**Issue by webmaster128** [iov-one/iov-core#618]
A BCP BlockId for Tendermint-based chains is an uppercase hex representation of a Tendermint [block hash](https://github.com/tendermint/tendermint/wiki/Block-Structure#block-hash).

Since we currently have no way to calculate the block hash from a block header, we need to do [an extra request](https://github.com/iov-one/iov-core/blob/v0.10.1/packages/iov-bns/src/bnsconnection.ts#L571-L572) for every event when watching block headers. This is something we want to avoid.

This is how to calculate:

* Check the [BlockHeader](https://github.com/tendermint/tendermint/blob/v0.27.0/types/block.go#L349-L380) structure
* [Hash the header](https://github.com/tendermint/tendermint/blob/v0.27.0/types/block.go#L404-L432)
* Using [cdcEncode](https://github.com/tendermint/tendermint/blob/v0.27.0/types/encoding_helper.go) and [merkle.SimpleHashFromByteSlices](https://github.com/tendermint/tendermint/blob/v0.27.0/crypto/merkle/simple_tree.go#L21-L34)
* "cbc" means codec and is implemented by [go-amino](https://github.com/tendermint/go-amino)

**Comment by ethanfrey** [comment]
This is needed for doing light-client proofs as well as saving the round-trip, so the importance will increase when we start looking to do proofs. Until we have other pieces ready or a burning need, it is too much work for the value IMO

**Comment by webmaster128** [comment]
This implementation may provide all we need: https://github.com/mappum/js-tendermint/blob/master/src/hash.js#L45

**Comment by willclarktech** [comment]
Link pinned to commit hash: https://github.com/nomic-io/js-tendermint/blob/9b83566/src/hash.js#L19-L39

**Comment by webmaster128** [comment]
There is an amino implementation in TypeScript, which might be helpful: https://github.com/node-a-team/ts-amino

**Comment by willclarktech** [comment]
@webmaster128 What are the acceptance criteria here? A function exported from `@iov/bns` that will take a block header and spit out a hash?

**Comment by webmaster128** [comment]
> What are the acceptance criteria here?

A correct implementation to replace this dummy block:

```
  public watchBlockHeaders(): Stream<BlockHeader> {
    // TODO: ID unavailable because
    // - we cannot trivially calculate it https://github.com/iov-one/iov-core/issues/618 and
    // - want to avoid an extra query to the node which causes issues when trying to send to a disconnected socket
    // Leave it a dummy as long as no application strictly requires the ID
    const dummyBlockId = "block ID not implemented for Tendermint" as BlockId;
```

**Comment by willclarktech** [comment]
@webmaster128 Our `Header` type is missing several fields that appear in these implementations, e.g. `version`. Is that just because we didn't need these fields so far or is there some other reason?

**Comment by webmaster128** [comment]
Parsing the `Header` in the Tendermint client is something that evolved along with the development of Tendermint. We probably had all the fields that were available in the initial version we connected to. But then the parser was only changed when:
1. We needed to access additional fields
2. Parsing the existing fields failed
For this reason it is very well possible that the parser is incomplete. Feel free to add the required fields.

If you find it helpful, consider doing https://github.com/iov-one/iov-core/issues/1249 first
